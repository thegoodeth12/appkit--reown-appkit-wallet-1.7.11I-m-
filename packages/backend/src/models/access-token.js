import Base from '@/models/base.js';
import User from '@/models/user.js';

class AccessToken extends Base {
  static tableName = 'access_tokens';

  static jsonSchema = {
    type: 'object',
    required: ['token', 'expiresIn'],

    properties: {
      id: { type: 'string', format: 'uuid' },
      userId: { type: 'string', format: 'uuid' },
      token: { type: 'string', minLength: 32 },
      samlSessionId: { type: ['string', 'null'] },
      expiresIn: { type: 'integer' },
      revokedAt: { type: ['string', 'null'], format: 'date-time' },
    },
  };

  static relationMappings = () => ({
    user: {
      relation: Base.BelongsToOneRelation,
      modelClass: User,
      join: {
        from: 'access_tokens.user_id',
        to: 'users.id',
      },
    },
  });

  async terminateRemoteSamlSession() {
    if (!this.samlSessionId) {
      return;
    }

    const user = await this.$relatedQuery('user');

    const firstIdentity = await user.$relatedQuery('identities').first();

    const samlAuthProvider = await firstIdentity
      .$relatedQuery('samlAuthProvider')
      .throwIfNotFound();

    const response = await samlAuthProvider.terminateRemoteSession(
      this.samlSessionId
    );

    return response;
  }

  async revoke() {
    const response = await this.$query().patch({
      revokedAt: new Date().toISOString(),
    });

    try {
      await this.terminateRemoteSamlSession();
    } catch (error) {
      // TODO: should it silently fail or not?
    }

    return response;
  }
}

export default AccessToken;
