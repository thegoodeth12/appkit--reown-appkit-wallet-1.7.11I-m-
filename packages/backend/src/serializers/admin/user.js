import userSerializer from '@/serializers/user.js';

const adminUserSerializer = (user) => {
  const userData = userSerializer(user);

  userData.acceptInvitationUrl = user.acceptInvitationUrl;

  return userData;
};

export default adminUserSerializer;
