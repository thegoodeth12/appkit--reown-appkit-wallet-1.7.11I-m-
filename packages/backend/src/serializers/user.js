import roleSerializer from '@/serializers/role.js';
import permissionSerializer from '@/serializers/permission.js';
import appConfig from '@/config/app.js';

const userSerializer = (user) => {
  let userData = {
    id: user.id,
    email: user.email,
    createdAt: user.createdAt.getTime(),
    updatedAt: user.updatedAt.getTime(),
    status: user.status,
    fullName: user.fullName,
  };

  if (user.role) {
    userData.role = roleSerializer(user.role);
  }

  if (user.permissions?.length > 0) {
    userData.permissions = user.permissions.map((permission) =>
      permissionSerializer(permission)
    );
  }

  if (appConfig.isCloud && user.trialExpiryDate) {
    userData.trialExpiryDate = user.trialExpiryDate;
  }

  return userData;
};

export default userSerializer;
