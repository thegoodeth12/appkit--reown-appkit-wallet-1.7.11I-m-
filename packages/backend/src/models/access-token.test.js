import { describe, it, expect, vi } from 'vitest';
import AccessToken from '@/models/access-token.js';
import User from '@/models/user.js';
import Base from '@/models/base.js';
import SamlAuthProvider from '@/models/saml-auth-provider.ee.js';
import { createAccessToken } from '@/factories/access-token.js';
import { createUser } from '@/factories/user.js';
import { createIdentity } from '@/factories/identity.js';

describe('AccessToken model', () => {
  it('tableName should return correct name', () => {
    expect(AccessToken.tableName).toBe('access_tokens');
  });

  it('jsonSchema should have correct validations', () => {
    expect(AccessToken.jsonSchema).toMatchSnapshot();
  });

  it('relationMappings should return correct associations', () => {
    const relationMappings = AccessToken.relationMappings();

    const expectedRelations = {
      user: {
        relation: Base.BelongsToOneRelation,
        modelClass: User,
        join: {
          from: 'access_tokens.user_id',
          to: 'users.id',
        },
      },
    };

    expect(relationMappings).toStrictEqual(expectedRelations);
  });

  it('revoke should set revokedAt and terminate remote SAML session', async () => {
    const accessToken = await createAccessToken();

    const terminateRemoteSamlSessionSpy = vi
      .spyOn(accessToken, 'terminateRemoteSamlSession')
      .mockImplementation(() => {});

    await accessToken.revoke();

    expect(terminateRemoteSamlSessionSpy).toHaveBeenCalledOnce();
    expect(accessToken.revokedAt).not.toBeUndefined();
  });

  describe('terminateRemoteSamlSession', () => {
    it('should terminate remote SAML session when exists', async () => {
      const user = await createUser();
      const accessToken = await createAccessToken({
        userId: user.id,
        samlSessionId: 'random-remote-session-id',
      });
      await createIdentity({ userId: user.id });

      const terminateRemoteSamlSessionSpy = vi
        .spyOn(SamlAuthProvider.prototype, 'terminateRemoteSession')
        .mockImplementation(() => {});

      await accessToken.terminateRemoteSamlSession();

      expect(terminateRemoteSamlSessionSpy).toHaveBeenCalledWith(
        accessToken.samlSessionId
      );
    });

    it(`should return undefined when remote SALM session doesn't exist`, async () => {
      const user = await createUser();
      const accessToken = await createAccessToken({ userId: user.id });
      await createIdentity({ userId: user.id });

      const terminateRemoteSamlSessionSpy = vi
        .spyOn(SamlAuthProvider.prototype, 'terminateRemoteSession')
        .mockImplementation(() => {});

      const expected = await accessToken.terminateRemoteSamlSession();

      expect(terminateRemoteSamlSessionSpy).not.toHaveBeenCalledOnce();
      expect(expected).toBeUndefined();
    });
  });
});
