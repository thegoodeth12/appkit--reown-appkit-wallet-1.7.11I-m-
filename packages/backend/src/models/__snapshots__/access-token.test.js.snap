// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AccessToken model > jsonSchema should have correct validations 1`] = `
{
  "properties": {
    "expiresIn": {
      "type": "integer",
    },
    "id": {
      "format": "uuid",
      "type": "string",
    },
    "revokedAt": {
      "format": "date-time",
      "type": [
        "string",
        "null",
      ],
    },
    "samlSessionId": {
      "type": [
        "string",
        "null",
      ],
    },
    "token": {
      "minLength": 32,
      "type": "string",
    },
    "userId": {
      "format": "uuid",
      "type": "string",
    },
  },
  "required": [
    "token",
    "expiresIn",
  ],
  "type": "object",
}
`;
